iree/testing/vulkan/iree-run-module-vulkan-gui-main.cc

```diff
@@ -0,0 +1,452 @@
+#include "iree/modules/hal/hal_module.h"
+#include "iree/tools/utils/vm_util.h"
+#include "iree/vm/api.h"
+#include "iree/vm/bytecode_module.h"
+
+ABSL_FLAG(std::string, module_file, "-",
+          "File containing the module to load that contains the entry "
+          "function. Defaults to stdin.");
+
+ABSL_FLAG(std::string, entry_function, "",
+          "Name of a function contained in the module specified by input_file "
+          "to run.");
+
+ABSL_FLAG(std::vector<std::string>, inputs, {},
+          "A comma-separated list of of input buffers of the format:"
+          "[shape]xtype=[value]\n"
+          "2x2xi32=1 2 3 4\n"
+          "Optionally, brackets may be used to separate the element values. "
+          "They are ignored by the parser.\n"
+          "2x2xi32=[[1 2][3 4]]\n"
+          "Due to the absence of repeated flags in absl, commas should not be "
+          "used to separate elements. They are reserved for separating input "
+          "values:\n"
+          "2x2xi32=[[1 2][3 4]], 1x2xf32=[[1 2]]");
+
+ABSL_FLAG(std::string, inputs_file, "",
+          "Provides a file for input shapes and optional values (see "
+          "ParseToVariantListFromFile in vm_util.h for details)");
+
+static VkAllocationCallbacks* g_Allocator = NULL;
+static VkInstance g_Instance = VK_NULL_HANDLE;
+static VkPhysicalDevice g_PhysicalDevice = VK_NULL_HANDLE;
+static VkDevice g_Device = VK_NULL_HANDLE;
+static uint32_t g_QueueFamily = (uint32_t)-1;
+static VkQueue g_Queue = VK_NULL_HANDLE;
+static VkPipelineCache g_PipelineCache = VK_NULL_HANDLE;
+static VkDescriptorPool g_DescriptorPool = VK_NULL_HANDLE;
+
+static ImGui_ImplVulkanH_Window g_MainWindowData;
+static uint32_t g_MinImageCount = 2;
+static bool g_SwapChainRebuild = false;
+static int g_SwapChainResizeWidth = 0;
+static int g_SwapChainResizeHeight = 0;
+
+namespace iree {
+namespace {
+
+void check_vk_result(VkResult err) {
+  if (err == 0) return;
+  IREE_LOG(FATAL) << "VkResult: " << err;
+}
+
+void CleanupVulkan() {
+  vkDestroyDescriptorPool(g_Device, g_DescriptorPool, g_Allocator);
+
+  vkDestroyDevice(g_Device, g_Allocator);
+  vkDestroyInstance(g_Instance, g_Allocator);
+}
+
+void CleanupVulkanWindow() {
+  ImGui_ImplVulkanH_DestroyWindow(g_Instance, g_Device, &g_MainWindowData,
+                                  g_Allocator);
+}
+
+StatusOr<std::string> GetModuleContentsFromFlags() {
+  auto module_file = absl::GetFlag(FLAGS_module_file);
+  std::string contents;
+  if (module_file == "-") {
+    contents = std::string{std::istreambuf_iterator<char>(std::cin),
+                           std::istreambuf_iterator<char>()};
+  } else {
+    IREE_ASSIGN_OR_RETURN(contents, file_io::GetFileContents(module_file));
+  }
+  return contents;
+}
+
+// Runs the current IREE bytecode module and renders its result to a window
+// using ImGui.
+Status ImGuiRender(iree_hal_device_t* device, iree_vm_context_t* context,
+                   iree_vm_function_t function) {
```

Comment:
Naming could be improved here, since this is doing more than rendering.

Games and other interactive programs will typically differentiate between "handle inputs", "update", and "draw"